File: 314__Binary_Tree_Vertical_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        record = collections.defaultdict(list)
        queue, result, min_i, max_i = [(0, root)], [], float("inf"), float("-inf")
        while queue:
            newq = []
            for i, node in queue:
                if node:
                    min_i = min(min_i, i)
                    max_i = max(max_i, i)
                    record[i].append(node.val)
                    newq.append((i - 1, node.left))
                    newq.append((i + 1, node.right))
            queue = newq
        for i in range(min_i, max_i + 1):
            result.append(record[i])
        return result

# More Simple Solution

class Solution(object):
    def verticalOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """

        idx_dict = collections.defaultdict(list)
        queue = [(root, 0)]
        for node, i in queue:
            if node:
                idx_dict[i].append(node.val)
                queue += (node.left, i - 1), (node.right, i + 1)

        return [idx_dict[idx] for idx in sorted(idx_dict.keys())]

File: Java/314__Binary_Tree_Vertical_Order_Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();
        Queue<TreeNode> queue = new LinkedList<>();
        HashMap<TreeNode, Integer> level_map = new HashMap<>();
        queue.add(root);
        level_map.put(root, 0);
        int min = 0;
        while (!queue.isEmpty()) {
            TreeNode node = queue.poll();
            int level = level_map.get(node);
            if (!map.containsKey(level)) {
                map.put(level, new ArrayList<>());
            }
            map.get(level).add(node.val);
            if (node.left != null) {
                queue.add(node.left);
                level_map.put(node.left, level - 1);
            }
            if (node.right != null) {
                queue.add(node.right);
                level_map.put(node.right, level + 1);
            }
            min = Math.min(min, level);
        }
        while (map.containsKey(min)) {
            result.add(map.get(min++));
        }
        return result;
    }

}
